test/utils/mocha.ts

```diff
@@ -1,11 +1,79 @@
+import deepcopy from 'deepcopy';
 import { expect } from 'chai';
 
+import FullStory from '../mocks/fullstory-recording';
 import { MockClass, Call } from '../mocks/mock';
 import { DataLayerDetail } from '../../src/event';
-import { DataLayerObserver, DataLayerConfig } from '../../src/observer';
+import { DataLayerObserver, DataLayerConfig, DataLayerRule } from '../../src/observer';
 import MonitorFactory from '../../src/monitor-factory';
 import { BuiltinOptions, OperatorFactory } from '../../src/factory';
 
+/**
+ * Gets or sets a global object. Use this rather than writing `globalThis as any`.
+ * @param key The key corresponding to the global object (e.g. FS in window['FS']).
+ * @param value Optional object that when provided will set the global's value
+ */
+export function global(key: string, value?: any): any {
+  if (value) {
+    (globalThis as any)[key] = value;
+  }
+
+  return (globalThis as any)[key];
+}
+
+/**
+ * A setup function that inits the global object (e.g. window) with desired expandos.
+ * Each expando will be deep copied to prevent cross-contamination between tests.
+ * Additionally, a FullStory mock will also be added to the `FS` expando.
+ * @param tuples List of expando tuples (e.g. ['FS', object])
+ */
+export function setupGlobals(tuples: [string, any][]) {
+  tuples.forEach((tuple) => {
+    expect(tuple[0]).to.be.ok;
+    expect(tuple[1]).to.be.ok;
```

Comment:
Wouldn't this be an overflow issue if only one tuple were passed into this function?

Reply:
It looks like you could define this as a type with shape:

beforeEach(() => setupGlobals([
    ['digitalData', basicDigitalData],
    ['_dlo_rules', rules],
  ]));

```
type dataLayerName =  "digitalData" | "s" | "dataLayer" | "utag";
type ruleName = "_dlo_rules";
type dataLayerDefinition = [dataLayerName, any];
type ruleDefinition = [ruleName, any];
setupGlobals(datalayer: dataLayerDefinition, rule: ruleDefinition)
```

I'm handing code so maybe there's an issue with this, but it could spare a potential bug down the line and is a maybe bit more descriptive.

Reply:
I made a few adjustments to the comments and function.  The main goal of the function was to reduce the overhead of setting up the `window`.  Since the `window` can take `any` value as an expando, that was the rationale to use less stringent type checking.  The function block doesn't make any type assumptions, it simply creates and assigns the expando (most importantly deep copying the provided value).  I did, however, beef up some of the expects inside the function block to be more informative on usage.

All that said, I'm wondering if this "setup" function should just add _all_ supported data layers.  I've skewed more to testing rulesets in isolation, but there's not really much overlap due to the different vendor's global objects.  What this approach does not address is future situations where I might want to set an arbitrary expando like `_dlo_destination`.  I could blend having the data layers auto-created and an optional that allows you to pass in additional key:value pairs for additional expandos.

Reply:
I was thinking the explicit types would add more clarity around the how data layers are named. I don't have a strong opinion, and I think your current implementation is fine.

> I'm wondering if this "setup" function should just add all supported data layers. I've skewed more to testing rulesets in isolation, but there's not really much overlap due to the different vendor's global objects.

I think you should set them up individually for each datalayer vendor test (as you are currently). This gets to my point around clarity of which data layer definition is tied to which vendor.